lib/craft/clock_bound/client.ex

```
@@ -0,0 +1,212 @@
+       when max_drift > 1_000_000_000 do
+    {:error, :max_drift_too_large}
+  end
+
+  defp compute_bound_at(real, monotonic_before, monotonic_after, clockbound_data) do
+    %{
+      as_of: as_of,
+      void_after: void_after,
+      clock_status: status,
+      max_drift: max_drift,
+      bound: bound
+    } = clockbound_data
+
+    # Validate whether the clockbound data read from shared memory is still valid at the time of the request by
+    # checking if the current monotonic time is lesser than the void_after time.
+    clock_status =
+      if status in [:synchronized, :free_running] and monotonic_after < void_after,
+        do: status,
+        else: :unknown
+
+    # Calculate the duration that has elapsed between the instant when the clockbound data were
+    # written to the shared memory segment (approximated by `as_of`), and the instant when the
+    # request to calculate the ClockErrorBound was actually requested (approximated by `monotonic`). This
+    # duration is used to compute the growth of the error bound due to local dispersion
+    # between polling chrony and now.
+    #
+    # Ideally, the current monotonic time should be greater than the time the clockbound data was written(as_of),
+    # but sometimes monotonic time is observed to be older by a handful of nanoseconds. To avoid this
+    # issue, we will introduce a small epsilon value (1ms) to account for clock precision.
+    # The cause of this behavior is not clear
+    causality_blur = as_of - 1000
+
+    duration =
+      cond do
+        # Happy path, clockbound data is older than the current monotonic time.
+        monotonic_before >= as_of -> monotonic_before - as_of
+        # Causality is "almost" broken but still within the epsilon range of clock precision as mentioned above.
+        # assume the monotonic time and the clockbound data updated time to be the same for this case (elapsed time = 0)
+        monotonic_before > causality_blur -> 0
```

the more i think about this fudge factor that the rust client authors built in to handwave away the behaviour they were observing of the monotonic clock on amazon's (possibly VM-obscured) hardware, the more it bothers me. on one hand if they think the clock is free running or possibly desynchronized, they just return that fact to the caller to have them decide, but on the other hand, if they observe bad behaviour of something that's meant to be guaranteed by the OS, they don't inform the caller and just sorta pretend that it didn't happen. that seems broken, or at the very least non-portable to systems that aren't as broken as amazon's.

if the monotonic clock is not abiding by its guarantees, i'd rather just throw the whole thing out and return an error indicating that the caller should try again. i don't want to hide the issue from the caller.

growing the error bound proportional to the time between when the SHM was written and when we read it totally makes sense.

so, in summary, i think that if the clock is not `:synchronized` or if `monotonic_before` is before `as_of`, we need to return an error. let's leave it up to the caller to decide to try again.

---

i agree thanks for the feedback.